lib/src/fixed_datetime_formatter.dart

```diff
@@ -0,0 +1,336 @@
+  /// [pattern].
+  DateTime decode(String formattedDateTime) {
+    return _decode(formattedDateTime, isUtc, true)!;
+  }
+
+  /// Parses [formattedDateTime] to a [DateTime] as specified by the [pattern].
+  ///
+  /// Parts of a [DateTime] which are not mentioned in the pattern default to a
+  /// value of zero for time parts and year, and a value of 1 for day and month.
+  ///
+  /// Returns the parsed value, or `null` if the [formattedDateTime] does not
+  /// match the [pattern].
+  DateTime? tryDecode(String formattedDateTime) {
+    return _decode(formattedDateTime, isUtc, false);
+  }
+
+  DateTime? _decode(
+    String formattedDateTime,
+    bool isUtc,
+    bool throwOnError,
+  ) {
+    var characters = formattedDateTime.codeUnits;
+    var year = 0;
+    var century = 0;
+    var decade = 0;
+    var month = 1;
+    var day = 1;
+    var hour = 0;
+    var minute = 0;
+    var second = 0;
+    var microsecond = 0;
+    for (var i = 0; i < _blocks.length; i++) {
+      var formatCharacter = _blocks.formatCharacters[i];
+      var number = _extractNumFromString(characters, i, throwOnError);
+      if (number != null) {
+        if (formatCharacter == _fractionSecondCode) {
+          // Special case, as we want fractional seconds to be the leading
+          // digits.
+          var numberLength = _blocks.ends[i] - _blocks.starts[i];
+          if (numberLength > 6) {
+            if (throwOnError) {
+              throw FormatException(
+                  'Fractional seconds can only be specified up to microseconds');
+            } else {
+              return null;
+            }
+          }
+          number *= _powersOfTen[6 - numberLength];
+        }
+        switch (formatCharacter) {
+          case _yearCode:
+            year = number;
+            break;
+          case _centuryCode:
+            century = number;
+            break;
+          case _decadeCode:
+            decade = number;
+            break;
+          case _monthCode:
+            month = number;
+            break;
+          case _dayCode:
+            day = number;
+            break;
+          case _hourCode:
+            hour = number;
+            break;
+          case _minuteCode:
+            minute = number;
+            break;
+          case _secondCode:
+            second = number;
+            break;
+          case _fractionSecondCode:
+            microsecond = number;
+            break;
+        }
+      } else {
+        return null;
+      }
+    }
+    var totalYear = year + 100 * century + 10 * decade;
+    if (isUtc) {
+      return DateTime.utc(
+        totalYear,
+        month,
+        day,
+        hour,
+        minute,
+        second,
+        0,
+        microsecond,
+      );
+    } else {
+      return DateTime(
+        totalYear,
+        month,
+        day,
+        hour,
+        minute,
+        second,
+        0,
+        microsecond,
+      );
+    }
+  }
+
+  int? _extractNumFromString(
```

Comment:
Name it `_extractNumberFromString` (avoid abbreviations, especially when the abbreviation suggests a different type, `num`).